packages/compass-workspaces/src/stores/workspaces.ts

```diff
@@ -843,28 +858,71 @@ export const openTabFromCurrent = (
   };
 };
 
-type CloseTabAction = { type: WorkspacesActions.CloseTab; atIndex: number };
+type DuplicateTabAction = {
+  type: WorkspacesActions.DuplicateTab;
+  atIndex: number;
+};
+
+export const duplicateTab = (atIndex: number): DuplicateTabAction => {
+  return {
+    type: WorkspacesActions.DuplicateTab,
+    atIndex,
+  };
+};
+
+async function confirmClosingTab() {
+  return await showConfirmation({
+    title: 'Are you sure you want to close the tab?',
+    description:
+      'The content of this tab has been modified. You will lose your changes if you close it.',
+    buttonText: 'Close tab',
+    variant: 'danger',
+    'data-testid': 'confirm-tab-close',
+  });
+}
+
+type CloseTabsAction = { type: WorkspacesActions.CloseTabs; tabIds: string[] };
 
 export const closeTab = (
   atIndex: number
-): WorkspacesThunkAction<Promise<void>, CloseTabAction> => {
+): WorkspacesThunkAction<Promise<void>, CloseTabsAction> => {
   return async (dispatch, getState) => {
-    const tab = getState().tabs[atIndex];
-    if (!canCloseTab(tab)) {
-      const confirmClose = await showConfirmation({
-        title: 'Are you sure you want to close the tab?',
-        description:
-          'The content of this tab has been modified. You will lose your changes if you close it.',
-        buttonText: 'Close tab',
-        variant: 'danger',
-        'data-testid': 'confirm-tab-close',
-      });
-      if (!confirmClose) {
-        return;
-      }
+    const { tabs } = getState();
+    const tab = tabs[atIndex];
+    if (canCloseTab(tab) || (await confirmClosingTab())) {
+      dispatch({ type: WorkspacesActions.CloseTabs, tabIds: [tab.id] });
+      cleanupRemovedTabs(tabs, getState().tabs);
     }
-    dispatch({ type: WorkspacesActions.CloseTab, atIndex });
-    cleanupLocalAppRegistryForTab(tab?.id);
+  };
+};
+
+export const closeAllOtherTabs = (
+  atIndex: number
+): WorkspacesThunkAction<Promise<void>, CloseTabsAction | SelectTabAction> => {
+  return async (dispatch, getState) => {
+    const { tabs } = getState();
+    const tabsToClose = await tabs.reduce(
+      async (prev: Promise<WorkspaceTab[]>, tab, tabIndex) => {
+        const tabsToClose = await prev;
+        if (tabIndex === atIndex) {
+          return tabsToClose; // Skip the tab which is not being closed
+        }
+        if (!canCloseTab(tab)) {
+          // Select the closing tab - to show the confirmation dialog in context
+          dispatch({ type: WorkspacesActions.SelectTab, atIndex: tabIndex });
```

Comment:
UI nit: as discussed, it makes sense to switch to the tab that's being closed during the process, but I found it a bit confusing that if you skip closing some of the tabs, you're not being returned back to the tab that you started with

Reply:
I've added an optional `activeTabId` to the `CloseTabs` action.